### PR TITLE
Expose ArgIterator

### PR DESCRIPTION
**Pull request by maryamariyan**
Exposing type ArgIterator which is already in System.Private.CoreLib

Fixes: https://github.com/dotnet/corefx/issues/32898

cc: @danmosemsft @safern @ViktorHofer

### CODE REVIEW

**Comment by ViktorHofer** [comment]
I know we just talked about it but I gave it a second thought and it would be nice if you could cross-compile for netcoreapp, remove the reflection and ifdef for netcoreapp: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs#L113

**Comment by stephentoub** [comment]
Can we add tests?

**Comment by danmosemsft** [comment]
If I understand @jkotas [comment](https://github.com/dotnet/corefx/issues/32898#issuecomment-434076736) this is only expected to work on Windows. Is that right @jashook ?

In which case tests will need to verify clean failure on Unix.

**Comment by jkotas** [comment]
>I know we just talked about it but I gave it a second thought and it would be nice if you could cross-compile for netcoreapp, remove the reflection and ifdef for netcoreapp: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs#L113

This test is of limited value. It is not exercising any new unique path in CoreCLR. I would just delete it instead of all this plumbing.

**Comment by jkotas** [comment]
> In which case tests will need to verify clean failure on Unix.

You may want to have `IsArgIteratorSupported` on PlatformDetection. I expect that the matrix where this is supported vs. not will get more complicated over time.

**Comment by jashook** [comment]
I agree with Jan, even today the matrix is not clear as Windows only is too permissive, for example Windows Arm32 is still missing work to have a functioning ArgIterator. Below is our current matrix.

| Windows  | Unix  | 
|---|---|
| <ul><li>- [x] x64</li></ul> | <ul><li>- [ ] x64 </li></ul> | 
| <ul><li>- [x] x86</li></ul> | <ul><li>- [ ] x86</li></ul>  |
| <ul><li>- [ ] armhf</li></ul> | <ul><li>- [ ] armhf </li></ul> |
|  | <ul><li>- [ ] armel</li></ul> | 
| <ul><li>- [x] arm64</li></ul> | <ul><li>- [ ] arm64</li></ul> |

**Comment by danmosemsft** [comment]
> You may want to have IsArgIteratorSupported on PlatformDetection.

That is useful for our tests -- do we need a detection mechanism for customers at runtime?

**Comment by jkotas** [comment]
> do we need a detection mechanism for customers at runtime?

This is a corner case API. I do not think we need an API to detect whether it is supported at runtime.

File: src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs [hunk 1]
```diff
@@ -109,13 +109,6 @@ public static IEnumerable<object[]> GetUninitializedObject_ByRefLikeType_TestDat
             if (!PlatformDetection.IsNetNative)
```
Comment (jkotas) [inline src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs]: You can delete the `if (!PlatformDetection.IsNetNative)` condition and the obsolete comment in front of it too while you are on it.

The test for `RuntimeArgumentHandle` should work just fine even on .NET Native.

File: src/System.Runtime/tests/System/Type/TypeTests.netcoreapp.cs [hunk 2]
```diff
@@ -220,6 +220,7 @@ public static IEnumerable<object[]> IsByRefLikeTestData
 
                 yield return new object[] { typeof(TypedReference), true };
                 yield return new object[] { typeof(RuntimeArgumentHandle), true };
+                yield return new object[] { typeof(ArgIterator), true };
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/Type/TypeTests.netcoreapp.cs]: sort first?

File: src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs [hunk 3]
```diff
@@ -66,6 +67,7 @@ public static partial class PlatformDetection
         public static bool IsNetNative { get { throw null; } }
         public static bool IsNonZeroLowerBoundArraySupported { get { throw null; } }
         public static bool IsNotIntMaxValueArrayIndexSupported { get { throw null; } }
+        public static bool IsNotArgIteratorSupported { get { throw null; } }
```
Comment (nil4) [inline src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs]: Wouldn't `IsArgIteratorNotSupported` read better? Currently it seems to say that something called `NotArgIterator` is supported?

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.cs [hunk 4]
```diff
@@ -37,6 +37,8 @@ public static partial class PlatformDetection
         public static bool IsNotArmProcess => !IsArmProcess;
         public static bool IsArm64Process => RuntimeInformation.ProcessArchitecture == Architecture.Arm64;
         public static bool IsNotArm64Process => !IsArm64Process;
+        public static bool IsArgIteratorSupported => IsWindows && (IsNotArmProcess || IsArm64Process);
+        public static bool IsNotArgIteratorSupported => !IsArgIteratorSupported;
```
Comment (nil4) [inline src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.cs]: Perhaps `IsArgIteratorNotSupported` instead?

**Comment by danmosemsft** [comment]
@maryamariyan is seems like you don't have a test that actually uses an ArgIterator. Only that test they serialize or are by ref. Is a test feasible? I have not used them myself. Seems like it ought to be.

**Pull request merged**